Introduce new Godot forum

winston-yallow opened the pull request:
This blog post introduces the new Godot forum and explains how we migrated away from the old Question2Answer platform

winston-yallow committed:
Introduce new Godot forum

JohnVeness commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+---
+title: "Introducing the new Godot Forum"
+excerpt: "We are excited to introduce a major improvement in terms of funding the development efforts of the Godot Engine."
```
Don't forget to change the excerpt!

JohnVeness commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+
+## Why did we replace the old platform?
+
+The old Q&A platform was based on Question2Answer, an open source project for Q&A sites. Unfortunately it is only infrequently updated and had some security issues we couldn't easily solve. This led to problems like bot registrations and spam.
```
```suggestion
The old Q&A platform was based on [Question2Answer](https://www.question2answer.org/), an open source project for Q&A sites. Unfortunately it is only infrequently updated and had some security issues we couldn't easily solve. This led to problems like bot registrations and spam.
```
Even though you're moving away from it, might be nice to give it a link.

JohnVeness commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+
+The old Q&A platform was based on Question2Answer, an open source project for Q&A sites. Unfortunately it is only infrequently updated and had some security issues we couldn't easily solve. This led to problems like bot registrations and spam.
+
+On top of that, we also wanted more forum features that Question2Answer simply couldn't provide. It served it's purpose well for quite some time, but we have simply outgrown it's capabilities. So it's time to give our community a new home!
```
```suggestion
On top of that, we also wanted more forum features that Question2Answer simply couldn't provide. It served its purpose well for quite some time, but we have simply outgrown its capabilities. So it's time to give our community a new home!
```

JohnVeness commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+
+## Introducing Discourse
+
+The new forum is based on [Discourse](https://www.discourse.org/), an open source forum software that can easily be adjusted via plugins and themes to suit our needs. For logins, the same SSO (single sign on) is used as for other Godot platforms (like the developer chat). This allows users to have one account accross all Godot websites.
```
```suggestion
The new forum is based on [Discourse](https://www.discourse.org/), an open source forum software that can easily be adjusted via plugins and themes to suit our needs. For logins, the same SSO (single sign on) is used as for other Godot platforms (like the developer chat). This allows users to have one account across all Godot websites.
```

JohnVeness commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+
+The new forum is based on [Discourse](https://www.discourse.org/), an open source forum software that can easily be adjusted via plugins and themes to suit our needs. For logins, the same SSO (single sign on) is used as for other Godot platforms (like the developer chat). This allows users to have one account accross all Godot websites.
+
+## Data Migration
```
```suggestion
## Data migration
```

JohnVeness commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+
+The old Q&A platform had over 150k questions, answers and comments stored in the database. 20% of those were hidden by our moderators for being spam or breaking the rules, so we ended up importing 120k items from the old database.
+
+What we did not import were the users. The old instance had lot's of old and unused accounts, and it would have been a lot of work to reconcile those users with the users of the SSO server. Instead, all imported posts are made by a system user, and the original username is shown at the top of the post. Since discourse does not support comments on posts like Question2Answer, we embed them directly ...
```
```suggestion
What we did not import were the users. The old instance had lots of old and unused accounts, and it would have been a lot of work to reconcile those users with the users of the SSO server. Instead, all imported posts are made by a system user, and the original username is shown at the top of the post. Since Discourse does not support comments on posts like Question2Answer, we embed them directly within the post and add some styling. This way we can represent all data from the old platform in a neat way:
```

JohnVeness commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+
+What we did not import were the users. The old instance had lot's of old and unused accounts, and it would have been a lot of work to reconcile those users with the users of the SSO server. Instead, all imported posts are made by a system user, and the original username is shown at the top of the post. Since discourse does not support comments on posts like Question2Answer, we embed them directly ...
+
+![imported post with note on top and comment below](/storage/blog/introducing-new-forum/example-import.webp)
```
```suggestion
![Imported post with note on top and comment below](/storage/blog/introducing-new-forum/example-import.webp)
```

JohnVeness commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+
+To keep existing links working, we forward any request from the old server to the new server. We then use the Discourse permalinks feature to map requests to `forum.godotengine.org/{some_old_id}` automatically to `forum.godotengine.org/t/{new_id}`.
+
+### External Images
```
```suggestion
### External images
```

JohnVeness commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+
+### External Images
+
+By default, discourse downloads external images on the server to prevent linkrot. However, this is not a good idea when importing a big amount of posts at the same time. Leaving this feature on would get the server blacklisted by imgur and other providers for requesting too many resources in a short amount of time. Therefore the feature was turned of for the duration of the import.
```
```suggestion
By default, Discourse downloads external images on the server to prevent linkrot. However, this is not a good idea when importing a big amount of posts at the same time. Leaving this feature on would get the server blacklisted by imgur and other providers for requesting too many resources in a short amount of time. Therefore the feature was turned off for the duration of the import.
```

JohnVeness commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+
+By default, discourse downloads external images on the server to prevent linkrot. However, this is not a good idea when importing a big amount of posts at the same time. Leaving this feature on would get the server blacklisted by imgur and other providers for requesting too many resources in a short amount of time. Therefore the feature was turned of for the duration of the import.
+
+### Markdown Adjustments
```
```suggestion
### Markdown adjustments
```

JohnVeness commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+
+### Markdown Adjustments
+
+Discourse uses markdown, HTML and BBCODE for the post content. Luckily, on the old Q&A platform we also used a plugin for markdown. While this made the posts somewhat compatible, there were still a few differences in how the markdown is interpreted.
```
```suggestion
Discourse uses Markdown, HTML and BBCODE for the post content. Luckily, on the old Q&A platform we also used a plugin for Markdown. While this made the posts somewhat compatible, there were still a few differences in how the Markdown is interpreted.
```

clayjohn commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+---
+title: "Introducing the new Godot Forum"
+excerpt: "We are excited to introduce a major improvement in terms of funding the development efforts of the Godot Engine."
```
```suggestion
excerpt: "We are finally unveiling the new Godot forum which replaces our Q & A platform while unlocking many new features."
```

JohnVeness replied:
I've had this debate on other sites, and I know people have different views, but is the new site "the Godot forum" or "the Godot forums"?

If it's "forum" (which I vote for), this should be "finally unveiling the new Godot forum which replaces"

clayjohn replied:
Updated based on your comment. I think you are right and "Forum" makes more sense

JohnVeness replied:
Don't forget the "s" on "replaces" :)

JohnVeness commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+
+Discourse uses markdown, HTML and BBCODE for the post content. Luckily, on the old Q&A platform we also used a plugin for markdown. While this made the posts somewhat compatible, there were still a few differences in how the markdown is interpreted.
+
+We ended up using a markdown parser to access the content in a tokenized form. This allows us to apply a bunch of adjustments:
```
```suggestion
We ended up using a Markdown parser to access the content in a tokenized form. This allows us to apply a bunch of adjustments:
```

JohnVeness commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+
+We ended up using a markdown parser to access the content in a tokenized form. This allows us to apply a bunch of adjustments:
+
+#### Code Blocks
```
```suggestion
#### Code blocks
```

clayjohn commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+
+## Why did we replace the old platform?
+
+The old Q&A platform was based on Question2Answer, an open source project for Q&A sites. Unfortunately it is only infrequently updated and had some security issues we couldn't easily solve. This led to problems like bot registrations and spam.
```
```suggestion
The old Q&A platform was based on Question2Answer, an open source project for Q&A sites. Unfortunately it is infrequently updated and had some security issues we couldn't easily solve. This led to problems like bot registrations and spam which took a lot of manual effort to manage.
```

Two changes here:
- A grammar fix (removing "only")
- highlighting the human cost of dealing with bots and spam

JohnVeness commented on collections/_article/introducing-new-forum.md:
````diff
@@ -0,0 +1,66 @@
+
+#### Code Blocks
+
+There are two ways to denote code blocks in markdown: either a block delimited by <code>```</code> or by indenting every line of the block by one tab (or four spaces). The old platform highlights both as code. Discourse only highlights backtick delimited blocks. So for all indent based code blocks, we remove the indentation and surround the blocks with backticks instead.
````
```suggestion
There are two ways to denote code blocks in Markdown: either a block delimited by <code>```</code> or by indenting every line of the block by one tab (or four spaces). The old platform highlights both as code. Discourse only highlights backtick delimited blocks. So for all indent based code blocks, we remove the indentation and surround the blocks with backticks instead.
```

JohnVeness commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+
+This ensures that all code blocks will be processed by `highlight.js` and will receive syntax highlighting if the language can be detected.
+
+#### Image Embeds and Links
```
```suggestion
#### Image embeds and links
```

JohnVeness commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+
+#### Image Embeds and Links
+
+The new markdown parser is a bit stricter about links/embeds that use the footnote syntax. They are only processed correctly if the line before is completely blank. This was easy to fix by simply adding a new line before every footnote. We also replace any references to the old server adress with references to the new server. This avoids unnecessary redirects.
```
```suggestion
The new Markdown parser is a bit stricter about links/embeds that use the footnote syntax. They are only processed correctly if the line before is completely blank. This was easy to fix by simply adding a new line before every footnote. We also replace any references to the old server address with references to the new server. This avoids unnecessary redirects.
```

clayjohn replied:
```suggestion
The new markdown parser is a bit stricter about links/embeds that use the footnote syntax. They are only processed correctly if the line before is completely blank. This was easy to fix by simply adding a new line before every footnote. We also replace any references to the old server address with references to the new server. This avoids unnecessary redirects.
```

JohnVeness commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+
+The new markdown parser is a bit stricter about links/embeds that use the footnote syntax. They are only processed correctly if the line before is completely blank. This was easy to fix by simply adding a new line before every footnote. We also replace any references to the old server adress with references to the new server. This avoids unnecessary redirects.
+
+![example post with working syntax highlighting and link](/storage/blog/introducing-new-forum/example-link-syntax.webp)
```
```suggestion
![Example post with working syntax highlighting and link](/storage/blog/introducing-new-forum/example-link-syntax.webp)
```

JohnVeness commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+
+![example post with working syntax highlighting and link](/storage/blog/introducing-new-forum/example-link-syntax.webp)
+
+## The Future
```
```suggestion
## The future
```

JohnVeness reviewed (commented):
A few nitpicks, but congrats on this overall!

clayjohn reviewed (commented):
Looks great! I'm very excited to share this with the community!

paddy-exe commented on collections/_article/introducing-new-forum.md:
```diff
@@ -0,0 +1,66 @@
+date: 2023-12-09 00:00:00
+---
+
+For the last couple of months, our Q&A platform was in readonly mode. Now it has been superceded by our new and more powerful forum! 🎉
```
```suggestion
For the last couple of months, our Q&A platform was in read-only mode. Now it has been superceded by our new and more powerful forum! 🎉
```

akien-mga replied:
And another one:
```suggestion
For the last couple of months, our Q&A platform was in read-only mode. Now it has been superseded by our new and more powerful forum! 🎉
```

winston-yallow commented:
Thank you all for the reviews :heart: 

I've added all your suggestions, set the publish date to 2023-12-08 and also squashed the commits. From my side this is good to be published :tada:

akien-mga commented:
Great work on the forum, and great write-up :)
